### ISSUE TITLE
PendingCmd panics on drop

### ISSUE THREAD

**Issue by brson** [tikv/tikv#4310]
It's not ok generally to panic in destructors because of the risk of double-panic ([see api guide](https://rust-lang-nursery.github.io/api-guidelines/dependability.html#destructors-never-fail-c-dtor-fail)), which causes an abort.

Here's the code:

```
impl Drop for PendingCmd {
    fn drop(&mut self) {
        if self.cb.is_some() {
            panic!(
                "callback of pending command at [index: {}, term: {}] is leak.",
                self.index, self.term
            );
        }
    }
}
```

What is the expected outcome for `cb` if the thread panicks? Is it ok to not call it in that situation? If that's ok then adding `!std::thread::panicking` to the condition in the dtor should be sufficient.

**Comment by brson** [comment]
Before trying to fix this I think we should get more clarity on the downsides of using `thread::panicking` in drop.

**Comment by brson** [comment]
I chatted a bit on #rust-libs about fallible destructor patterns today, and came away thinking we should go ahead and use thread::panicking to avoid the double-panic.

That is, if it is acceptable to not call that callback on panic. Hopefully someone will think about that during review.

Generally though, I think the best pattern for fallible destructors is to have a close method that returns a Result, and a drop method that verifies that close was called. I sketched out how this looks [here](https://gist.github.com/brson/4ab0526baab38972bccdfb60ef7a9bd9), though I must say I've never seen code with the dual `close()`/`destroy()` methods like that.
